pkg/object/httpserver/routers/spec.go

```diff
@@ -0,0 +1,322 @@
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package routers
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+
+	"github.com/megaease/easegress/pkg/logger"
+	"github.com/megaease/easegress/pkg/util/ipfilter"
+	"github.com/megaease/easegress/pkg/util/stringtool"
+)
+
+type Rules []*Rule
+
+type Paths []*Path
+
+type Rule struct {
+	// NOTICE: If the field is a pointer, it must have `omitempty` in tag `json`
+	// when it has `omitempty` in tag `jsonschema`.
+	// Otherwise it will output null value, which is invalid in json schema (the type is object).
+	// the original reason is the jsonscheme(genjs) has not support multiple types.
+	// Reference: https://github.com/alecthomas/jsonschema/issues/30
+	// In the future if we have the scenario where we need marshal the field, but omitempty
+	// in the schema, we are suppose to support multiple types on our own.
+	IPFilterSpec *ipfilter.Spec `json:"ipFilter,omitempty" jsonschema:"omitempty"`
+	Host         string         `json:"host" jsonschema:"omitempty"`
+	HostRegexp   string         `json:"hostRegexp" jsonschema:"omitempty,format=regexp"`
+	Paths        Paths          `json:"paths" jsonschema:"omitempty"`
+
+	ipFilter *ipfilter.IPFilter
+	hostRE   *regexp.Regexp
+}
+
+type Path struct {
+	IPFilterSpec      *ipfilter.Spec `json:"ipFilter,omitempty" jsonschema:"omitempty"`
+	Path              string         `json:"path,omitempty" jsonschema:"omitempty,pattern=^/"`
+	PathPrefix        string         `json:"pathPrefix,omitempty" jsonschema:"omitempty,pattern=^/"`
+	PathRegexp        string         `json:"pathRegexp,omitempty" jsonschema:"omitempty,format=regexp"`
+	RewriteTarget     string         `json:"rewriteTarget" jsonschema:"omitempty"`
+	Methods           []string       `json:"methods,omitempty" jsonschema:"omitempty,uniqueItems=true,format=httpmethod-array"`
+	Backend           string         `json:"backend" jsonschema:"required"`
+	ClientMaxBodySize int64          `json:"clientMaxBodySize" jsonschema:"omitempty"`
+	Headers           Headers        `json:"headers" jsonschema:"omitempty"`
+	Queries           Queries        `json:"queries,omitempty" jsonschema:"omitempty"`
+	MatchAllHeader    bool           `json:"matchAllHeader" jsonschema:"omitempty"`
+	MatchAllQuery     bool           `json:"matchAllQuery" jsonschema:"omitempty"`
+
+	ipFilter             *ipfilter.IPFilter
+	method               MethodType
+	cacheable, matchable bool
```

Comment:
propose not to use these two fields, they bring little performance benefit but make the code hard to read and error prone.